### PR TITLE
Implement `list_root_keys` to the `AdminKeyValueStore`

### LINKED ISSUE #3085: Introduce the `get_root_keys` to the `AdminKeyValueStore`
We are currently using the `root_key` in the `AdminKeyValueStore`. This is a formalization of the notion of partition keys in databases.

What is missing is the ability to get the list of root keys. That is we would like to have
```rust
fn get_root_keys(&self) -> Result<Vec<Vec<u8>>, Self::Error>
```

There are two use cases for that:
* Accessing to the list of chains of a validator.
* Migrating data from one storage to another.

There are obstacles to that:
* For DynamoDb and ScyllaDb, we could access the root keys easily since they are partition keys.
* The `RocksDb` does not have a notion of partition key.
  - Instead, we simply concatenate the keys like `root_key + key`.
  - This design was chosen because having one RocksDB instance per root key would lead to an overflow in the number of opened files.
  - One idea would be to add some entry indicating the creation of a new root key when doing a `clone_with_root_key`. This would make the `clone_with_root_key` async but that is fine.
* The support in storage service would also be somewhat difficult. But a priori possible.

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

We need to have the feature of accessing to the list of root keys in order to have mutation function and accessing to the list of chain ids of a storage.

Fixes #3085 

## Proposal

The implementation causes some problems:
* For ScyllaDb the implementation is very easy since the root_key is used as a partition key.
* For DynamoDb we can implement the root_key as a partition key, but DynamoDb forbids iterating and determining all the partition keys. So, we need to keep track of the keys.
* For RocksDb / StorageService / IndexedDb we need to keep track of the list of root_keys.
* For RocksDb, this led to a simplification since the edge case of having a key of the form `[255, ..., 255]` disappears. This corrects a problem and a test is added to detect it.
* The result of the `list_root_keys` will not be the same on different storage. If storage has been created with `fn create`, some keys were written but later deleted then in DynamoDb, RocksDb, storage-service, IndexedDb the root key will show up as existing while in ScyllaDb, the root key will not be visible.

The writing of the root key occurs when a `write_batch` is done.

## Test Plan

One test has been added for this feature.

## Release Plan

No impact on the TestNet / DevNet. It can follow the normal release plan.

## Links

None.

### CODE REVIEW

File: linera-views/src/backends/dynamo_db.rs [hunk 1]
```diff
@@ -92,11 +92,19 @@ async fn get_config_internal(
     }
 }
 
+/// The DynamoDb forbids the iteration over the partition keys.
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:95]: Ok, thanks.

File: linera-views/src/backends/dynamo_db.rs [hunk 2]
```diff
@@ -92,11 +92,19 @@ async fn get_config_internal(
...
+/// Therefore we use a special partition_key named [1] for storing
+/// the root keys. For normal root_keys, we simply put a 0 in front
+/// therefore no intersection is possible.
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:98]: Yes, corrected.

File: linera-views/src/backends/dynamo_db.rs [hunk 3]
```diff
@@ -159,9 +167,9 @@ const MAX_TRANSACT_WRITE_ITEM_SIZE: usize = 100;
 
 /// Keys of length 0 are not allowed, so we extend by having a prefix on start
 fn extend_root_key(root_key: &[u8]) -> Vec<u8> {
-    let mut vec = vec![0];
-    vec.extend(root_key);
-    vec
+    let mut start_key = vec![0];
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:170]: Ok.

File: linera-views/src/backends/dynamo_db.rs [hunk 4]
```diff
@@ -159,9 +167,9 @@ const MAX_TRANSACT_WRITE_ITEM_SIZE: usize = 100;
...
+    start_key.extend(root_key);
+    start_key
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:172]: I am actually against that. I first used the term `extended_root_key` but it is not great. The `start_key` is short enough to be easy to use. It is also used in the backends in a fairly uniform way with the same meaning when it is used.

The inconvenience of `prefix` and `key_prefix` is that they are used for searching purposes. I prefer not to have a collision of the meaning of names.

File: linera-views/src/backends/dynamo_db.rs [hunk 5]
```diff
@@ -308,7 +314,7 @@ impl TransactionBuilder {
         value: Vec<u8>,
         store: &DynamoDbStoreInternal,
     ) -> Result<(), DynamoDbStoreInternalError> {
-        let transact = store.build_put_transact(&self.root_key, key, value)?;
+        let transact = store.build_put_transact(&self.start_key, key, value)?;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs]: Ok, I used `transaction` in the DynamoDb code.

File: linera-client/src/client_options.rs [hunk 6]
```diff
@@ -964,6 +964,14 @@ pub enum DatabaseToolCommand {
         #[arg(long = "storage")]
         storage_config: String,
     },
+
+    /// List the rot keys of the database
```
Comment (afck) [inline linera-client/src/client_options.rs:968]:
```suggestion
    /// List the root keys of the database
```

File: linera-client/src/storage.rs [hunk 7]
```diff
@@ -554,6 +554,32 @@ impl StoreConfig {
             }
         }
     }
+
+    /// Lists all root keys of the storage
+    pub async fn list_root_keys(self) -> Result<Vec<Vec<u8>>, ViewError> {
+        match self {
+            StoreConfig::Memory(_, _) => Err(ViewError::StoreError {
+                backend: "memory".to_string(),
+                error: "list_all is not supported for the memory storage".to_string(),
```
Comment (afck) [inline linera-client/src/storage.rs:563]:
```suggestion
                error: "list_all_root_keys is not supported for the memory storage".to_string(),
```

File: linera-storage-service/proto/key_value_store.proto [hunk 8]
```diff
@@ -139,6 +139,24 @@ message ReplyListAll {
...
+}
+
+
+message RequestInsertRootKey {
+  bytes namespace = 1;
+  bytes root_key = 2;
+}
+
+message ReplyInsertRootKey {
+}
```
Comment (afck) [inline linera-storage-service/proto/key_value_store.proto:157]: (We could use `google.protobuf.Empty` instead of all those empty types. Not sure if that's preferable, but we do it in `rpc.proto`.)

File: linera-views/tests/admin_tests.rs [hunk 9]
```diff
@@ -7,27 +7,53 @@ use linera_views::dynamo_db::DynamoDbStore;
...
+    root_key_admin_test::<RocksDbStore>().await;
+}
+
+#[cfg(with_dynamodb)]
+#[tokio::test]
+async fn root_key_admin_test_dynamo_db() {
+    root_key_admin_test::<DynamoDbStore>().await;
+}
+
+#[cfg(with_scylladb)]
```
Comment (afck) [inline linera-views/tests/admin_tests.rs:55]: Might be more readable to use `test_case` for this.

File: linera-views/src/store.rs [hunk 10]
```diff
@@ -174,11 +174,18 @@ pub trait LocalAdminKeyValueStore: WithError + Sized {
...
 
     /// Takes a connection and creates a new one with a different `root_key`.
-    fn clone_with_root_key(&self, root_key: &[u8]) -> Result<Self, Self::Error>;
+    async fn clone_with_root_key(&self, root_key: &[u8]) -> Result<Self, Self::Error>;
 
     /// Obtains the list of existing namespaces.
     async fn list_all(config: &Self::Config) -> Result<Vec<String>, Self::Error>;
 
+    /// Lists the root keys of the namespace.
+    /// It is possible that some root_keys have no keys.
```
Comment (afck) [inline linera-views/src/store.rs:183]: Do you mean "some namespaces have no root keys"?

**Review by afck — COMMENTED** [review]
Sorry, I think I need some more explanations (or better: code comments, or pointers where to find them) before I can review this. I'm confused how the keys are structured and represented in the different databases. E.g. I thought the first byte was there to distinguish between actual key-value pairs, and markers for the existence of namespaces; in what way does that relate to root keys at all, and why is it changing now?

File: linera-client/src/client_options.rs (hunk 6 above)
Comment (MathieuDutSik) [reply linera-client/src/client_options.rs:968]: Thank you.

File: linera-client/src/storage.rs (hunk 7 above)
Comment (MathieuDutSik) [reply linera-client/src/storage.rs:563]: Thanks.

File: linera-views/src/store.rs (hunk 10 above)
Comment (MathieuDutSik) [reply linera-views/src/store.rs:183]: No, I really mean that.
What happens is that for each key, there is at least one root key.
But for some root key, there is no corresponding key.
* For ScyllaDb, we will have that the insertion is done via `(root_ley, key)` which lead to the fact that for each key we have a matching root_key.
* For others the root key has to be stored as a key. That creation is done just at `fn connect` or `fn clone_with_root_key`. So, the creation itself creates root keys. But the key is not yet written.

File: linera-views/src/test_utils/mod.rs [hunk 11]
```diff
@@ -700,3 +703,70 @@ where
...
+        }
+    }
+
+    let read_root_keys = S::list_root_keys(&config, &namespace)
+        .await
+        .expect("read_root_keys");
+    let set_root_keys = root_keys.iter().cloned().collect::<HashSet<_>>();
+    for read_root_key in &read_root_keys {
+        assert!(set_root_keys.contains(read_root_key));
+    }
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/test_utils/mod.rs:755]: No.
For a database like ScyllaDb, the root keys would be inserted only if there is a corresponding key. This is why there is a `size_select` that could be 0.

File: linera-views/src/backends/scylla_db.rs [hunk 12]
```diff
@@ -715,6 +718,33 @@ impl AdminKeyValueStore for ScyllaDbStoreInternal {
...
+        // See TODO(#1069).
+        let query = format!("SELECT root_key FROM kv.{} ALLOW FILTERING", namespace);
+
+        // Execute the query
+        let rows = session.query_iter(query, &[]).await?;
+        let mut rows = rows.rows_stream::<(Vec<u8>,)>()?;
+        let mut root_keys = HashSet::new();
+        while let Some(row) = rows.next().await {
+            let (root_key,) = row?;
+            let root_key = root_key[1..].to_vec();
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/backends/scylla_db.rs:742]: In ScyllaDb, a key has to hava non-zero length. Therefore, we have a function `get_big_root_key` that adds a 0 at the first entry.

File: linera-views/src/backends/rocks_db.rs [hunk 13]
```diff
@@ -193,52 +188,28 @@ impl RocksDbStoreExecutor {
...
-                    if let Excluded(upper_bound) = get_upper_bound(&key_prefix) {
-                        let mut full_key1 = self.root_key.to_vec();
-                        full_key1.extend(key_prefix);
-                        let mut full_key2 = self.root_key.to_vec();
-                        full_key2.extend(upper_bound);
-                        inner_batch.delete_range(&full_key1, &full_key2);
-                    }
+                    let mut full_key1 = self.start_key.to_vec();
+                    full_key1.extend(&key_prefix);
+                    let full_key2 = get_upper_bound_option(&full_key1).unwrap();
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/backends/rocks_db.rs:211]: Yes, exactly.
Added, documentation.

File: linera-views/tests/admin_tests.rs (hunk 9 above)
Comment (MathieuDutSik) [reply linera-views/tests/admin_tests.rs:55]: Yes, thank you.

File: linera-storage-service/proto/key_value_store.proto (hunk 8 above)
Comment (MathieuDutSik) [reply linera-storage-service/proto/key_value_store.proto:157]: Thank you, that is great.

File: linera-views/src/backends/rocks_db.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-views/src/backends/rocks_db.rs:212]: You may remove the comment and replace `.unwrap()` by `.expect("the first entry cannot be 255");`

File: linera-views/src/backends/rocks_db.rs [hunk 14]
```diff
@@ -96,7 +91,7 @@ fn check_key_size(key: &[u8]) -> Result<(), RocksDbStoreInternalError> {
 #[derive(Clone)]
 struct RocksDbStoreExecutor {
     db: Arc<DB>,
-    root_key: Vec<u8>,
+    start_key: Vec<u8>,
```
Comment (ma2bd) [inline linera-views/src/backends/rocks_db.rs:94]: what is a `start_key`? Seems like we're missing field documentation here.

File: linera-views/src/backends/rocks_db.rs [hunk 15]
```diff
@@ -425,19 +401,30 @@ impl AdminKeyValueStore for RocksDbStoreInternal {
...
     }
 
-    fn clone_with_root_key(&self, root_key: &[u8]) -> Result<Self, RocksDbStoreInternalError> {
+    async fn clone_with_root_key(
+        &self,
+        root_key: &[u8],
+    ) -> Result<Self, RocksDbStoreInternalError> {
         let mut store = self.clone();
-        store.executor.root_key = root_key.to_vec();
+        store.executor.start_key = vec![1];
```
Comment (ma2bd) [inline linera-views/src/backends/rocks_db.rs:421]: I feel that we should define two global constants instead of `0` and `1`.

Comment (ma2bd) [reply linera-views/src/backends/rocks_db.rs:421]: The notion of "key space" or "domain" comes to mind.

File: linera-views/src/backends/scylla_db.rs [hunk 16]
```diff
@@ -715,6 +718,33 @@ impl AdminKeyValueStore for ScyllaDbStoreInternal {
...
+    ) -> Result<Vec<Vec<u8>>, ScyllaDbStoreInternalError> {
+        Self::check_namespace(namespace)?;
+        let session = SessionBuilder::new()
+            .known_node(config.uri.as_str())
+            .build()
+            .boxed()
+            .await?;
+        // Create a table if it doesn't exist
+        // The schema appears too complicated for non-trivial reasons.
+        // See TODO(#1069).
```
Comment (ma2bd) [inline linera-views/src/backends/scylla_db.rs:733]: #1069 is already closed. Do we need to the TODO?

File: linera-client/src/client_options.rs [hunk 17]
```diff
@@ -964,6 +964,14 @@ pub enum DatabaseToolCommand {
         #[arg(long = "storage")]
         storage_config: String,
     },
+
+    /// List the root keys of the database
+    #[command(name = "list_root_keys")]
+    ListRootKeys {
```
Comment (ma2bd) [inline linera-client/src/client_options.rs:970]: In `linera-client`, we know that the "root keys" are (basically) chain ids. So shouldn't we list the chain ids instead?

Comment (ma2bd) [reply linera-client/src/client_options.rs:970]: And similarly we should list ApplicationId, BlobId, etc as such.

Comment (ma2bd) [reply linera-client/src/client_options.rs:970]: Btw you may need to add conversion function in `linera-storage` for this.

File: linera-views/src/backends/rocks_db.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/backends/rocks_db.rs:212]: Ok.

File: linera-views/src/backends/rocks_db.rs (hunk 14 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/rocks_db.rs:94]: It is a fact that what is used externally as `root_key` does not translate exactly like that in the specific code. Building it everytime it is needed would cost a little, which is why it is better to store the key that is going to be used.

We need to add something and it varies according to the context. This happens for `RocksDb`, `DynamoDb`, `IndexDb`, `LineraStorageService`. Description could be added but there would not be unifying explanation as this is all case specific.

About the name, I thought `effective_root_key`, `extended_root_key`. But they are quite long. Which is why I liked `start_key` which is short and reasonable in what it expresses.

File: linera-views/src/backends/rocks_db.rs (hunk 15 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/rocks_db.rs:421]: Ok, except for the second where I prefer `STORE_ROOT_KEYS_PREFIX`.

File: linera-views/src/backends/scylla_db.rs (hunk 16 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/scylla_db.rs:733]: Yes, thank you.

File: linera-client/src/client_options.rs (hunk 17 above)
Comment (MathieuDutSik) [reply linera-client/src/client_options.rs:970]: Yes, this is why I wanted to do it in a subsequent PR.

File: linera-views/src/backends/scylla_db.rs [hunk 18]
```diff
@@ -715,6 +718,30 @@ impl AdminKeyValueStore for ScyllaDbStoreInternal {
...
+            .known_node(config.uri.as_str())
+            .build()
+            .boxed()
+            .await?;
+        let query = format!("SELECT root_key FROM kv.{} ALLOW FILTERING", namespace);
+
+        // Execute the query
+        let rows = session.query_iter(query, &[]).await?;
+        let mut rows = rows.rows_stream::<(Vec<u8>,)>()?;
+        let mut root_keys = HashSet::new();
```
Comment (afck) [inline linera-views/src/backends/scylla_db.rs:736]: If we use a `BTreeSet`, it will return the keys sorted, rather than in a random order.

File: linera-views/src/backends/dynamo_db.rs [hunk 19]
```diff
@@ -91,11 +92,17 @@ async fn get_config_internal(
     }
 }
 
+/// DynamoDb forbids the iteration over the partition keys.
+/// Therefore we use a special partition_key named [1] for storing
+/// the root keys. For normal root_keys, we simply put a 0 in front
```
Comment (afck) [inline linera-views/src/backends/dynamo_db.rs:97]:
```suggestion
/// DynamoDB forbids the iteration over the partition keys.
/// Therefore we use a special partition key named `[1]` for storing
/// the root keys. For normal root keys, we simply put a `[0]` in front
```

File: linera-views/src/backends/dynamo_db.rs [hunk 20]
```diff
@@ -352,28 +370,34 @@ impl AdminKeyValueStore for DynamoDbStoreInternal {
...
+        store.write_root_key(root_key).await?;
+        Ok(store)
     }
 
-    fn clone_with_root_key(&self, root_key: &[u8]) -> Result<Self, DynamoDbStoreInternalError> {
+    async fn clone_with_root_key(
+        &self,
+        root_key: &[u8],
+    ) -> Result<Self, DynamoDbStoreInternalError> {
+        self.write_root_key(root_key).await?;
```
Comment (afck) [inline linera-views/src/backends/dynamo_db.rs:389]: Does that mean `DbStorage::load_chain` will now write a root key?

E.g. if we handle a certificate only to realize during execution that it's invalid, it would still write that chain state view's root key to storage?

File: linera-views/src/backends/rocks_db.rs [hunk 21]
```diff
@@ -425,19 +405,30 @@ impl AdminKeyValueStore for RocksDbStoreInternal {
...
+    async fn clone_with_root_key(
+        &self,
+        root_key: &[u8],
+    ) -> Result<Self, RocksDbStoreInternalError> {
         let mut store = self.clone();
-        store.executor.root_key = root_key.to_vec();
+        store.executor.start_key = STORED_ROOT_KEYS_PREFIX.to_vec();
+        let mut batch = Batch::new();
+        batch.put_key_value_bytes(root_key.to_vec(), vec![]);
+        store.write_batch(batch).await?;
```
Comment (afck) [inline linera-views/src/backends/rocks_db.rs:428]: (See ScyllaDB: Not sure we should write here.)

File: linera-views/src/backends/dynamo_db.rs (hunk 19 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:97]: Ok.

File: linera-views/src/backends/dynamo_db.rs (hunk 20 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:389]: Yes.
That is the design with the root keys. The running of the system would still be deterministic.

File: linera-views/src/backends/rocks_db.rs (hunk 21 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/rocks_db.rs:428]: Can you precise?
RocksDD does not have any sort of partition key. So, we need to implement them manually.

File: linera-views/src/backends/scylla_db.rs (hunk 18 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/scylla_db.rs]: That is true, that does not really matter, but we can do it that way.

File: linera-views/src/backends/dynamo_db.rs (hunk 20 above)
Comment (afck) [reply linera-views/src/backends/dynamo_db.rs:389]: But we're calling `load_chain` quite early in the validation process: You could spam a validator with completely invalid certificates with random bogus chain IDs, and it would create a database entry for each of these IDs.

Comment (ma2bd) [reply linera-views/src/backends/dynamo_db.rs:389]: Yes, this sounds much better to delay the registration of a root view to the first `write_batch`.

Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs]: Ok, done.

File: linera-views/src/backends/dynamo_db.rs [hunk 22]
```diff
@@ -972,6 +958,25 @@ impl DirectWritableKeyValueStore for DynamoDbStoreInternal {
     type Batch = SimpleUnorderedBatch;
 
     async fn write_batch(&self, batch: Self::Batch) -> Result<(), DynamoDbStoreInternalError> {
+        let write_root_key = {
+            let mut root_key_written = self.root_key_written.lock().expect("root_key_written should not be poisoned");
+            if !*root_key_written {
+                *root_key_written = false;
```
Comment (afck) [inline linera-views/src/backends/dynamo_db.rs:968]: Did you mean to set it to `true`?

File: linera-storage-service/src/common.rs [hunk 23]
```diff
@@ -28,6 +28,8 @@ pub enum KeyTag {
     Key = MIN_VIEW_TAG,
     /// Prefix for the storage of existence or not of the namespaces.
     Namespace,
+    /// Prefix for the root key
+    RootKey,
```
Comment (ma2bd) [inline linera-storage-service/src/common.rs:32]: The first variant should probably be called `RootView`.
* "Prefix for the storage of the keys of the map" -> "Key prefix for the storage of the root views"
* etc

File: linera-views/src/backends/dynamo_db.rs [hunk 24]
```diff
@@ -91,11 +96,17 @@ async fn get_config_internal(
...
+/// DynamoDB forbids the iteration over the partition keys.
+/// Therefore we use a special partition key named `[1]` for storing
+/// the root keys. For normal root keys, we simply put a `[0]` in
+/// front therefore no intersection is possible.
+const PARTITION_KEY_ROOT_KEY: &[u8] = &[1];
+
 /// The attribute name of the partition key.
 const PARTITION_ATTRIBUTE: &str = "item_partition";
 
 /// A root key being used for testing existence of tables
```
Comment (ma2bd) [inline linera-views/src/backends/dynamo_db.rs:108]: Is this a "key" or a "key prefix"?

File: linera-storage-service/src/client.rs [hunk 25]
```diff
@@ -240,8 +240,20 @@ impl WritableKeyValueStore for ServiceStoreClientInternal {
             return Ok(());
         }
         let mut statements = Vec::new();
+        let mut full_key = self.start_key.clone();
+        full_key[0] = KeyTag::RootKey as u8;
+        let operation = Operation::Put(KeyValue {
+            key: full_key,
+            value: vec![],
+        });
```
Comment (ma2bd) [inline linera-storage-service/src/client.rs:248]: It may not be ideal to overwrite this key over and over. (E.g. it might evict other keys faster from any cache)
Would it be possible to only schedule one such write per connect?

Comment (MathieuDutSik) [reply linera-storage-service/src/client.rs:248]: Yes, we can do that.
Only write at the first, this requires a `Arc<Mutex<bool>>` which I did for DynamoDb, but not for others as I thought it was needed for DynamoDb.

File: linera-views/src/backends/rocks_db.rs [hunk 26]
```diff
@@ -193,55 +192,35 @@ impl RocksDbStoreExecutor {
...
+                    full_key1.extend(&key_prefix);
+                    let full_key2 =
+                        get_upper_bound_option(&full_key1).expect("the first entry cannot be 255");
+                    inner_batch.delete_range(&full_key1, &full_key2);
                 }
             }
         }
+        let mut full_key = self.start_key.to_vec();
+        full_key[0] = STORED_ROOT_KEYS_PREFIX;
+        inner_batch.put(&full_key, vec![]);
```
Comment (ma2bd) [inline linera-views/src/backends/rocks_db.rs:223]: Same. Do we have to write this every time?

Comment (ma2bd) [reply linera-views/src/backends/rocks_db.rs:223]: That one is called "keys_prefix" lol

File: linera-views/src/backends/dynamo_db.rs (hunk 22 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:968]: Yes, thank you.

File: linera-views/src/backends/dynamo_db.rs (hunk 24 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:108]: It is a key.
The partition key coming from root keys start with 0, the partition key storing the root keys is exactly `[1]`.

File: linera-views/src/backends/rocks_db.rs (hunk 26 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/rocks_db.rs]: Corrected. Now we write every time only for IndexedDB.

File: linera-storage-service/src/client.rs [hunk 27]
```diff
@@ -54,14 +57,18 @@ const MAX_KEY_SIZE: usize = 1000000;
...
+//   to indicate the existence of a root key.
 #[derive(Clone)]
 pub struct ServiceStoreClientInternal {
     channel: Channel,
     semaphore: Option<Arc<Semaphore>>,
     max_stream_queries: usize,
-    namespace: Vec<u8>,
+    prefix_len: usize,
     start_key: Vec<u8>,
+    root_key_written: Arc<Mutex<bool>>,
```
Comment (afck) [inline linera-storage-service/src/client.rs:71]: Maybe an `AtomicBool` would be lower overhead. Also, you could do `.fetch_or(true)` below to set and get the value atomically.

File: linera-storage-service/src/client.rs [hunk 28]
```diff
@@ -308,7 +337,7 @@ impl ServiceStoreClientInternal {
     }
 
     fn namespace_as_vec(namespace: &str) -> Result<Vec<u8>, ServiceStoreError> {
-        let mut key = vec![KeyTag::Key as u8];
+        let mut key = vec![];
         bcs::serialize_into(&mut key, namespace)?;
         Ok(key)
```
Comment (afck) [inline linera-storage-service/src/client.rs:342]: I think that's equivalent to:
```suggestion
        Ok(bcs::to_bytes(namespace)?)
```

Alternatively, maybe this should take a `u8` as an argument to prepend to the namespace?

File: linera-views/src/backends/dynamo_db.rs [hunk 29]
```diff
@@ -320,7 +327,8 @@ pub struct DynamoDbStoreInternal {
     namespace: String,
     semaphore: Option<Arc<Semaphore>>,
     max_stream_queries: usize,
-    root_key: Vec<u8>,
+    start_key: Vec<u8>,
+    root_key_written: Arc<Mutex<bool>>,
```
Comment (afck) [inline linera-views/src/backends/dynamo_db.rs:331]: Same here: Should probably be `AtomicBool`.

File: linera-views/src/backends/dynamo_db.rs [hunk 30]
```diff
@@ -400,6 +411,24 @@ impl AdminKeyValueStore for DynamoDbStoreInternal {
...
     }
 
+    async fn list_root_keys(
+        config: &Self::Config,
+        namespace: &str,
+    ) -> Result<Vec<Vec<u8>>, DynamoDbStoreInternalError> {
+        let mut store = Self::connect(config, namespace, &[]).await?;
+        store.start_key = PARTITION_KEY_ROOT_KEY.to_vec();
+
+        let key_prefix = &[0];
```
Comment (afck) [inline linera-views/src/backends/dynamo_db.rs:421]: Should we use `EMPTY_ROOT_KEY` here?

File: linera-views/src/backends/rocks_db.rs [hunk 31]
```diff
@@ -193,55 +196,41 @@ impl RocksDbStoreExecutor {
...
-                if get_upper_bound(key_prefix) == Bound::Unbounded {
-                    for short_key in self.find_keys_by_prefix_internal(key_prefix.to_vec())? {
-                        let mut full_key = self.root_key.clone();
-                        full_key.extend(key_prefix);
-                        full_key.extend(short_key);
-                        keys.push(full_key);
-                    }
-                }
-            }
-        }
```
Comment (afck) [inline linera-views/src/backends/rocks_db.rs:214]: Does this change the semantics of a batch?

In general, if we have a `DeletePrefix` in a batch, and later a `Put` in that prefix, I guess we _do_ want that to be written, right? (So the old code violated that.)

File: linera-views/src/backends/rocks_db.rs [hunk 32]
```diff
@@ -254,6 +243,7 @@ pub struct RocksDbStoreInternal {
     _path_with_guard: PathWithGuard,
     max_stream_queries: usize,
     spawn_mode: RocksDbSpawnMode,
+    root_key_written: Arc<Mutex<bool>>,
```
Comment (afck) [inline linera-views/src/backends/rocks_db.rs:246]:
```suggestion
    root_key_written: Arc<AtomicBool>,
```

File: linera-views/src/backends/rocks_db.rs [hunk 33]
```diff
@@ -402,9 +397,24 @@ impl WritableKeyValueStore for RocksDbStoreInternal {
...
+                .root_key_written
+                .lock()
+                .expect("root_key_written should not be poisoned");
+            if !*root_key_written {
+                *root_key_written = true;
+                true
+            } else {
+                false
+            }
+        };
```
Comment (afck) [inline linera-views/src/backends/rocks_db.rs:411]:
```suggestion
        let write_root_key = !root_key_written.fetch_or(true);
```

File: linera-views/src/test_utils/mod.rs [hunk 34]
```diff
@@ -700,3 +703,67 @@ where
             .expect("A successful deletion");
     }
 }
+
+/// Testing of the root key functionality
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/test_utils/mod.rs:707]: I prefer the second.

File: linera-views/src/backends/rocks_db.rs (hunk 33 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/rocks_db.rs:411]: That is great. But the writing is
```
let write_root_key = !self.root_key_written.fetch_or(true, Ordering::SeqCst);
```

which is still shorter than the original code.

File: linera-views/src/backends/rocks_db.rs (hunk 32 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/rocks_db.rs:246]: Yes, that is simpler.

File: linera-views/src/backends/dynamo_db.rs (hunk 29 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:331]: Yes, that works.

File: linera-storage-service/src/client.rs (hunk 28 above)
Comment (MathieuDutSik) [reply linera-storage-service/src/client.rs:342]: Thanks, this lead to the elimination of `namespace_as_vec`.

File: linera-storage-service/src/client.rs (hunk 27 above)
Comment (MathieuDutSik) [reply linera-storage-service/src/client.rs:71]: Yes, thanks.

File: linera-storage-service/src/common.rs (hunk 23 above)
Comment (MathieuDutSik) [reply linera-storage-service/src/common.rs]: I am against putting the word "view". It is true that storage can be used for views but that is not the only usage.
I understand that the `MIN_VIEW_TAG` can give the wrong impression and so I removed it.

File: linera-storage-service/src/client.rs (hunk 27 above)
Comment (afck) [reply linera-storage-service/src/client.rs:72]: It's still a mutex in this file.

Comment (ma2bd) [reply linera-storage-service/src/client.rs:72]: I was going to ask why :)

File: linera-views/src/backends/dynamo_db.rs (hunk 30 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/dynamo_db.rs:424]: Yes, we can.

File: linera-views/src/backends/rocks_db.rs (hunk 31 above)
Comment (MathieuDutSik) [reply linera-views/src/backends/rocks_db.rs:214]: Yes, the semantics changes. I added a test that is passed by the new code and failed by the old.
